HW4_Voskoboinikov/ultimate_protein_tools.py

```diff
@@ -0,0 +1,691 @@
+
+
+def get_protein_rnas_number(seq: int, **_) -> int:
+    """
+    Get number of all possible RNA's for a given protein.
+
+    Arguments:
+    - seq (str): seq to be checked
+
+    Return:
+    - rnas_num (int): number of possible RNA's for seq
+    """
+
+    rnas_num = 1
+    for amino_acid in seq:
+        rnas_num *= len(RNA_AA_TABLE[amino_acid])
+    return rnas_num
+
+
+def check_all_upper(codon: str) -> bool:
+    """
+    Checks whether all letters in colon are upper
+
+    Arguments:
+    - codon (str): codon to be checked
+
+    Return:
+    - check_upper (bool): if all letters are uppercase
+    """
+
+    check_upper = True
+    for letter in (set(codon)):
+        check_upper = letter.isupper() and check_upper
+    return check_upper
+
+
+def get_frameshift_proteins(seq: int,
+                            check_if_user_conscious: bool = False,
+                            is_stop_codon_termination_enabled: bool = False,
+                            **_) -> dict:
+    """
+    Returns list of all possible proteins from all possible frames in peptide.
+
+    WARNING: can be computationally intensive on longer sequences,
+    will NOT start unless check_if_user_conscious is True!
+    
+    Arguments:
+    - seq (str): seq to be checked
+    - check_if_user_conscious (bool): checks user's consciousness. Default False
+    - is_stop_codon_termination_enabled (bool): terminate translation when reached stop-codon. Default False.
+
+    Return:
+    - dict: dict of lists of all possible frames proteins:
+    {frame_0: ['protein_seqs'], frame_1: ['protein_seqs'], frame_2: ['protein_seqs']}
+    """
+
+    if check_if_user_conscious:
+        frameshift_dct = {'frame_0': [seq]}  # set current seq as frame_0 (protein from not-shifted frame)
+        rnas = get_protein_rnas(seq, check_if_user_conscious=check_if_user_conscious)
+        for frame_number in [1, 2]:
+            frames_list = []
+            for rna in rnas:
+                frame = ''
+                for i in range(frame_number, len(rna) - (frame_number + 1), 3):  # set frame-dependent range to iterate
+                    frame_codon = rna[i:i + 3]  # extract codon
+                    if not check_all_upper(frame_codon):  # check if all letters in codon uppercase
+                        frame_codon = frame_codon.lower()  # if not change all to lowercase
+                    frame += RNA_CODON_TABLE[frame_codon]
+                    if is_stop_codon_termination_enabled and RNA_CODON_TABLE[
+                        frame_codon] == '*':  # stop writing if meet stop-codon
+                        break
+                frames_list.append(frame)  # append frame to frames list
+            frameshift_dct[f'frame_{frame_number}'] = list(set(frames_list))  # clean duplicates and write to dict
+        return frameshift_dct
+    return "You don't fucking know what you're doing!"  # politely ask user to reconsider their actions
+
+
+def get_length_of_protein(seq: str, **_) -> int:
```

Comment:
```suggestion
def get_protein_length(seq: str) -> int:
```

Ну, конечно, тут ошибиться так-то негде:)
Я спокойно принимал эту функцию в других работах, потому что может не хватить фантазии / времени и так есть способ добить чем то простеньким. Ладно бы хотя бы оно правильно считало длину и в однобуквенном и в трехбуквенном коде. Но так-то это совсем примитивно)
Опять же, мы это принимали, но тут вы написали в 2 раза больше функций чем надо. Мы всё равно все в любом случае проверяем (а еще перепроверяем). В первую очередь это отнимает скорость и качество проверки у ваших коллег)). И теперь еще эта функция выделяется как одна из ваших лучших функций. Я не согласен. Почти все (если не все) ваши функции лучше чем эта, даже не смотря на какие-то синтаксические недочеты. 
Я не буду сейчас ставить за неё 0, хотя, если честно, захотелось )